### PR TITLE
Add post on cranelift backend

### PR DESCRIPTION
**Pull request by jyn514**
After the fact I realized you can just use `rustup-toolchain-install-master` and now I feel silly :( but at least this is a guide on how to build in-tree. Let me know if I should just use the instructions in https://github.com/bjorn3/rustc_codegen_cranelift#building instead.

r? @bjorn3

### CODE REVIEW

**Comment by bjorn3** [comment]
> After the fact I realized you can just use rustup-toolchain-install-master and now I feel silly

Only `./x.py check` builds it by default. `./x.py build` has it disabled by default and so is it on CI. Either way I haven't setup `./x.py dist` support at all.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:37]: This post should probably be blocked on merging that PR.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Should this also mention how to build it from the main repo? That is a lot quicker.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Yeah, that's what I meant by 'I feel silly'. I'll move building in-tree to a separate section and use the `prepare.sh` for the rest of the article.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: @bjorn3 is there a way to build cg_clif without also running the tests?

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: actually, test.sh didn't work either:
```
[BUILD] sysroot
error: could not execute process `/home/joshua/rustc_codegen_cranelift/build_sysroot/../target/debug/cg_clif_build_sysroot -vV` (never executed)

Caused by:
  No such file or directory (os error 2)

real	0m0.043s
user	0m0.036s
sys	0m0.010s
```

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Not at the moment. I can add something.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: >  test.sh didn't work either:

I think this is because I have `CARGO_TARGET_DIR` set; I unset it in the parent but I guess running a bash shell set it again.

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: I will unset it in `build_sysroot.sh`.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Unsetting `CARGO_TARGET_DIR` didn't help. I think there's some cache invalidation issue, maybe?

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: bjorn3/rustc_codegen_cranelift@324e63de289e249481445a399fcbcad62b7ab71d unsets `CARGO_TARGET_DIR` where necessary.

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Just saw your previous comment. Did you build in debug mode? (not passing `--release` to `test.sh`)

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:27]: Well, either way it works with `build --release && build_sysroot.sh` directly, so probably no need to worry about it.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jonas-schievink) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:15]:
```suggestion
only performing very few optimizations. This is a great fit for debug builds, and the hope is
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:63]: How much is the difference in incremental recompilation time? cg_clif doesn't support internal parallelism until parallel rustc lands. This means that cg_llvm can normally utilize the cpu cores more effectively when there are few crates that can be built at a given time. In incremental mode, it likely only needs to recompile a single cgu, so there won't be any internal parallelism for cg_llvm too.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:63]: Does `cg_llvm` support internal parallelism? Is that parallism in LLVM itself?

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: cg_clif was actually slower, not faster :/

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: cg_llvm optimizes codegen units in parallel.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: Still slower if I only change main.rs:
```
$ CARGO_TARGET_DIR=cg_clif ../rustc_codegen_cranelift/cargo.sh build
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
$ cargo +nightly-2020-10-31 build
   Compiling cargo v0.50.0 (/home/joshua/cargo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
```
Maybe it's because I compiled cg_clif locally but the nightly version comes from CI and does LTO?

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: serde_derive shows up a lot in the profile. I think it is simply due to the fact that serde_derive is not very optimized in this case.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:11]:
```suggestion
codegen backend for the Rust compiler. The existing backend is LLVM, which is very
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:15]: I suggest either removing the "only" or changing "very" to "a":

```suggestion
performing very few optimizations. This is a great fit for debug builds, and the hope is
```

```suggestion
only performing a few optimizations. This is a great fit for debug builds, and the hope is
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: Probably https://github.com/serde-rs/json/issues/727 then.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:20]:
```suggestion
There has been a [Major Change Proposal][MCP] open for some time for making
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:24]:
```suggestion
`cg_clif` is not yet distributed with `rustup`, but this means you can now
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:40]:
```suggestion
Now, we can start using it in a project. For demonstration
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:90]:
```suggestion
CPU cores more effectively when there are few crates that can be built at a
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:97]:
```suggestion
`cg_clif` is because the Rust compiler takes a very long time to build.
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: That issue is about compiletime of crates using serde_json due to large amounts of ir being generated. This issue is serde_derive being slow. It takes ~14% of the compilation time for libcargo.

**Review by camelid — COMMENTED** [review]
Some copy-editing :)

Also, your filename is incorrect: it should be "using-rustc_codegen_cranelift" not "using-rustc_cranelift_codegen"

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: Oops :laughing:

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:20]: Hmm, I say 'an em-c-p', which has a vowel at the start. Do you pronounce it differently?

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:40]: 'in a project' sounds like it's a dependency. Maybe 'to compile a project'?

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: For cg_llvm it is <3% of the compilation time spent in serde_derive for libcargo.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:40]: Sure, something like that sounds fine. The "our own" part is what sounds funny, because it implies that you can use it for some projects beforehand, but not personal projects.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (camelid) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:20]: I pronounce it the same. What you wrote renders as:

> There has been an [Major Change Proposal]().

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:20]: oh oops :laughing:

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:91]: In this case it turned out to be serde_derive being slower at runtime when compiled using cg_clif than cg_llvm.

Comment (jyn514) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: Updated it with the info about serde_derive.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:37]:
```suggestion
$ ./build.sh
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:48]:
```suggestion
$ ../rustc_codegen_cranelift/build/cargo.sh build
```

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:81]:
```suggestion
$ ../rustc_codegen_cranelift/build/build.sh
```

**Review by bjorn3 — CHANGES_REQUESTED** [review]
Changed build instructions for https://github.com/bjorn3/rustc_codegen_cranelift/pull/1104

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md:102]: It just got merged.

File: posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md
Comment (jyn514) [inline posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: This is inconsistent with `build.sh` above, which isn't in the `build/` subdirectory. Which did you mean? @bjorn3

Comment (bjorn3) [reply posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md]: Ah, this should be `build/cargo.sh build`.

**Comment by jyn514** [comment]
This is ready to go I think :)

**Review by bjorn3 — APPROVED** [review]
Thanks a lot for writing this!

**Comment by jyn514** [comment]
@XAMPPRocky I think this is ready to be merged :)

**Comment by jyn514** [comment]
ping @rust-lang/core, I had someone asking how to do this just now. It would be great to get this merged, or if it needs work I can do that too.

**Review by Mark-Simulacrum — APPROVED** [review]

**Comment by Mark-Simulacrum** [comment]
Happy to merge with date bumped to November 16th (or whatever is current when you do so), please ping me on Zulip so I can quickly hit the button.

It would also be good to migrate the cg_clif repository into the rust-lang org soon, but I don't think that's a blocker for publishing this.

**Comment by jyn514** [comment]
@Mark-Simulacrum updated :)

**Pull request merged**
